data-prepper-plugins/failures-common/src/main/java/org/opensearch/dataprepper/plugins/dlq/s3/S3DlqWriter.java

```diff
@@ -0,0 +1,141 @@
+    static final String S3_DLQ_REQUEST_SUCCESS = "dlqS3RequestSuccess";
+    static final String S3_DLQ_REQUEST_FAILED = "dlqS3RequestFailed";
+    static final String S3_DLQ_REQUEST_LATENCY = "dlqS3RequestLatency";
+    static final String S3_DLQ_REQUEST_SIZE_BYTES = "dlqS3RequestSizeBytes";
+    private static final String KEY_NAME_FORMAT = "dlq-v%s-%s-%s-%s-%s";
+    private static final String FULL_KEY_FORMAT = "%s/%s";
+
+    private final S3Client s3Client;
+    private final String bucket;
+    private final String keyPathPrefix;
+    private final ObjectMapper objectMapper;
+
+    private final Counter dlqS3RecordsSuccessCounter;
+    private final Counter dlqS3RecordsFailedCounter;
+    private final Counter dlqS3RequestSuccessCounter;
+    private final Counter dlqS3RequestFailedCounter;
+    private final Timer dlqS3RequestTimer;
+    private final DistributionSummary dlqS3RequestSizeBytesSummary;
+
+    S3DlqWriter(final S3DlqWriterConfig s3DlqWriterConfig, final ObjectMapper objectMapper, final PluginMetrics pluginMetrics) {
+        dlqS3RecordsSuccessCounter = pluginMetrics.counter(S3_DLQ_RECORDS_SUCCESS);
+        dlqS3RecordsFailedCounter = pluginMetrics.counter(S3_DLQ_RECORDS_FAILED);
+        dlqS3RequestSuccessCounter = pluginMetrics.counter(S3_DLQ_REQUEST_SUCCESS);
+        dlqS3RequestFailedCounter = pluginMetrics.counter(S3_DLQ_REQUEST_FAILED);
+        dlqS3RequestTimer = pluginMetrics.timer(S3_DLQ_REQUEST_LATENCY);
+        dlqS3RequestSizeBytesSummary = pluginMetrics.summary(S3_DLQ_REQUEST_SIZE_BYTES);
+
+        this.s3Client = s3DlqWriterConfig.getS3Client();
+        Objects.requireNonNull(s3DlqWriterConfig.getBucket());
+        this.bucket = s3DlqWriterConfig.getBucket();
+        this.keyPathPrefix = s3DlqWriterConfig.getKeyPathPrefix();
+        this.objectMapper = objectMapper;
+    }
+
+    @Override
+    public void write(final List<DlqObject> dlqObjects, final String pipelineName, final String pluginId) throws IOException {
+        try {
+            doWrite(dlqObjects, pipelineName, pluginId);
+            dlqS3RequestSuccessCounter.increment();
+            dlqS3RecordsSuccessCounter.increment(dlqObjects.size());
+        } catch (final Exception e) {
+            dlqS3RequestFailedCounter.increment();
+            dlqS3RecordsFailedCounter.increment(dlqObjects.size());
+            throw e;
+        }
+    }
+
+    private void doWrite(final List<DlqObject> dlqObjects, final String pipelineName, final String pluginId) throws IOException {
+        final PutObjectRequest putObjectRequest = PutObjectRequest.builder()
+            .bucket(bucket)
+            .key(buildKey(pipelineName, pluginId))
+            .build();
+
+        final String content = deserialize(dlqObjects);
+
+        final PutObjectResponse response = timedPutObject(putObjectRequest, content);
+
+        if (!response.sdkHttpResponse().isSuccessful()) {
+            throw new IOException(String.format(
```

Comment:
Adding metrics based on the response code here will be useful. You can do it in a different PR as an improvement.

Reply:
#2440 <- tracking here